### PR TITLE
Bazel build of pytorch

### PR DESCRIPTION
**Pull request by werkt**

### CODE REVIEW

**Comment by albanD** [comment]
Just adding reviewers to make sure it is looked at by someone. Feel free to remove yourself if someone else is already looking into it.

File: tools/setup_helpers/generate_code.py [hunk 1]
```diff
@@ -34,27 +34,35 @@ def generate_code(ninja_global=None,
...
+    autograd_gen_dir = '%s/autograd/generated' % install_dir
+    jit_gen_dir = '%s/jit/generated' % install_dir
     for d in (autograd_gen_dir, jit_gen_dir):
         if not os.path.exists(d):
             os.makedirs(d)
+    runfiles_dir = os.environ["RUNFILES_DIR"]
+    if runfiles_dir:
+        data_dir = os.path.join(runfiles_dir, 'pytorch')
+    else:
+        data_dir = ""
```
[the remark this answers is not in the document]
Comment (werkt) [reply tools/setup_helpers/generate_code.py]: Considered this. ternaries in python aren't my favorite, but no problem.

File: third_party/fbgemm.BUILD [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        "src/FbgemmFloat16ConvertAvx512.cc",
+        "src/FbgemmI64.cc",
+        "src/FbgemmI8Spmdm.cc",
+        "src/GenerateKernelU8S8S32ACC16.cc",
+        "src/GenerateKernelU8S8S32ACC16Avx512.cc",
+        "src/GenerateKernelU8S8S32ACC16Avx512VNNI.cc",
+        "src/GenerateKernelU8S8S32ACC32.cc",
+        "src/GenerateKernelU8S8S32ACC32Avx512.cc",
+        "src/GenerateKernelU8S8S32ACC32Avx512VNNI.cc",
+        "src/GroupwiseConvAcc32Avx2.cc",
```
[the remark this answers is not in the document]
Comment (werkt) [reply third_party/fbgemm.BUILD]: These are a part of the generic srcs for the cmake targets. They must use... JIT?

File: third_party/fbgemm.BUILD [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        "src/UtilsAvx2.cc",
+        # Private headers
+        "src/FbgemmFP16Common.h",
+        "src/FbgemmFP16UKernelsAvx2.h",
+        "src/FbgemmI8Depthwise2DAvx2-inl.h",
+        "src/FbgemmI8DepthwiseAvx2-inl.h",
+        "src/MaskAvx2.h",
+        "src/OptimizedKernelsAvx2.h",
+        "src/TransposeUtils.h",
+        "src/TransposeUtilsAvx2.h",
```
[the remark this answers is not in the document]
Comment (werkt) [reply third_party/fbgemm.BUILD]: They can, but I was seeking to emulate what fbgemm does as a part of its build.

File: third_party/fbgemm.BUILD [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        ":fbgemm_avx512",
+        ":fbgemm_src_headers",
+        "@asmjit",
+        "@cpuinfo",
+    ],
+)
+
+cc_library(
+    name = "fbgemm_avx2",
+    srcs = [
```
[the remark this answers is not in the document]
Comment (werkt) [reply third_party/fbgemm.BUILD]: Fixed

File: third_party/cpuinfo.BUILD [hunk 5]
```diff
@@ -0,0 +1,62 @@
+# UATC LLC (c) 2019
```
Comment (malfet) [inline third_party/cpuinfo.BUILD]: @werkt can you please either remove copyright headers from .BUILD files or expand it to full BSD-style headers?

Comment (werkt) [reply third_party/cpuinfo.BUILD]: Done. I missed this in my conversion, after looking for it and being confused that I hadn't found any of our copyrights.

**Review by malfet — APPROVED** [review]

File: WORKSPACE [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+new_local_repository(
+    name = "pybind11",
+    build_file = "//third_party:pybind11.BUILD",
+    path = "third_party/pybind11",
+)
+
+http_archive(
+    name = "python3",
+    build_file = "//third_party:python3.BUILD",
+    urls = ["https://artifacts.int.uberatc.com/artifactory/auto-bazel/38dd863d274a6f6e37264ffce5e0da76e70affa06451245167b1fcd3b93cbec3/python3.tar.gz"],
```
Comment (driazati) [inline WORKSPACE]: These look like uber internal resources, is this supposed to be accessible from open source?

Comment (werkt) [reply WORKSPACE]: Too right. I'll get this removed.

File: .circleci/config.yml [hunk 7]
```diff
@@ -2308,6 +2308,20 @@ workflows:
                 - /ci-all\/.*/
           build_environment: "pytorch-linux-xenial-py3-clang5-android-ndk-r19c-arm-v8a-build"
           docker_image: "308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:f990c76a-a798-42bb-852f-5be5006f8026"
+      - pytorch_linux_build:
+          name: pytorch_linux_xenial_py3_6_bazel_build
+          requires:
+            - setup
+          build_environment: "pytorch-linux-xenial-py3.6-bazel-build"
+          docker_image: "308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-bazel:f990c76a-a798-42bb-852f-5be5006f8026"
```
[the remark this answers is not in the document]
Comment (seemethere) [reply .circleci/config.yml]: Maybe it's possible that we do this without code generation and just write an entry in `verbatim-sources` for the bazel build

[the remark this answers is not in the document]
Comment (werkt) [reply .circleci/config.yml]: That's fine, but I assumed that the bare execution environment would be unsuitable without a container initialization.

**Comment by nicolov** [comment]
Thanks a lot for your efforts upstreaming this. Are you interested in patches/feedback in the case where pytorch is included as a third-party repository, or are you planning to first merge a working standalone build, and then make changes?

**Comment by werkt** [comment]
> Thanks a lot for your efforts upstreaming this. Are you interested in patches/feedback in the case where pytorch is included as a third-party repository, or are you planning to first merge a working standalone build, and then make changes?

Merge a working copy that prevents breakages. We (Uber) are doing this because we have the same use case, and will be working towards novel third party inclusion utilities after it has landed. This is just to get a foot in the door.

**Comment by nicolov** [comment]
> Merge a working copy that prevents breakages.

Perfect, makes sense. Are you planning to include CUDA support in the first round?

**Comment by werkt** [comment]
No, cuda is coming later. We use this internally for it, but the FB torch team was inboard for cpu first. Hermetic cuda (the goal) requires a toolchain, which we haven't formalized for release. You can see some snippets of our cuda definitions in these BUILD files for now

**Comment by nicolov** [comment]
> No, cuda is coming later.

Cool, thanks. I'm migrating our internal BUILD files for pytorch to this, and was wondering whether to wait or not. I'm guessing you can import the CUDA-on-clang toolchain from TensorFlow?

**Comment by werkt** [comment]
> > No, cuda is coming later.
> 
> Cool, thanks. I'm migrating our internal BUILD files for pytorch to this, and was wondering whether to wait or not. I'm guessing you can import the CUDA-on-clang toolchain from TensorFlow?

Negative. The clang-primary cuda compilation does not work, nvcc must be used as a frontend for cuda compilations with pytorch. We investigated that early on and could not make it work.

**Comment by nicolov** [comment]
> We investigated that early on and could not make it work.

What types of errors were you getting? I do have it working with clang 9 and CUDA 10.

**Comment by werkt** [comment]
> > We investigated that early on and could not make it work.
> 
> What types of errors were you getting? I do have it working with clang 9 and CUDA 10.

"have it working" can you elaborate? You are currently building all pytorch cuda sources with a device target with clang as the exec'd compiler frontend, *not* with clang as the host cc?

**Comment by nicolov** [comment]
> "have it working" can you elaborate?

We don't use nvcc at all, just the clang PTX backend directly. IIRC there's a TensorFlow configuration that does the same.

**Comment by werkt** [comment]
> We don't use nvcc at all, just the clang PTX backend directly. IIRC there's a TensorFlow configuration that does the same.

Taking this conversation offline.

**Comment by nicolov** [comment]
For reference, these are the (small) changes I have to make this work as an external repo (also supports remote execution - that's why I had to add a bunch of Python imports to the `gen` script). https://gist.github.com/nicolov/09af3c1a8c7dcf971822daf8f2dc9caf

File: BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,2492 @@
...
+        "c10/cuda/*.h",
+        "c10/cuda/impl/*.h",
+        "c10/macros/*.h",
+        "c10/util/*.h",
+    ]) + [
+        "c10/macros/cmake_macros.h",
+        "c10/cuda/impl/cuda_cmake_macros.h",
+    ],
+    deps = [
+        "@com_github_gflags_gflags//:gflags",
```
Comment (nicolov) [inline BUILD.bazel:63]: I wonder how you can get by with this. `c10_headers` in linked in one of the .sos that have linkstatic, and my build complains about duplicate flags definitions.

**Review by seemethere — APPROVED** [review]

**Comment by seemethere** [comment]
@werkt sorry to do this to you but it appears as though we need a rebase.

**Comment by werkt** [comment]
> @werkt sorry to do this to you but it appears as though we need a rebase.

Not a problem. The end of that test clause changed twice in the last hour :(

**Pull request closed without merging**